opentracing-impl-java8/src/main/java/io/opentracing/propagation/TextFormatExtractorImpl.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.propagation;
+
+import java.util.Map;
+
+import io.opentracing.Tracer;
+
+final class TextFormatExtractorImpl implements TextFormatExtractor {
+
+    /** How the trace state is serialised and can be identified separately to baggage during deserialisation,
+     * for the tracing implementation being used, is contextual to the extraction process.
+     */
+    interface TraceState {
+        /** Returns true if the key+value belongs in traceState, otherwise it is presumed to be baggage. */
+        boolean isTraceState(String key, Object value);
+    }
+
+    private final TraceState traceState;
+
+    TextFormatExtractorImpl(TraceState traceState) {
+        this.traceState = traceState;
+    }
+
+    @Override
+    public Tracer.SpanBuilder join(Map carrier) {
+        AbstractTracer.SpanBuilderImpl builder = new AbstractTracer.SpanBuilderImpl();
+        for (Map.Entry<String, String> entry : ((Map<String, String>) carrier).entrySet()) {
+            if (traceState.isTraceState(entry.getKey(), entry.getValue())) {
+                builder.traceState.put(entry.getKey(), entry.getValue());
+            } else {
```

Comment:
Can we not impose that only traceState or baggage attributes exist in the carrier supplied?

I don't know if this is within the scope of the OT Specification, but it would work given that TextFormatExtractorImpl is going to be delegated to from a specific Extractor implementation.

It also make sense in that filtering makes more sense as part of the delegation chain, rather than supplied via context into TextFormatExtractorImpl and run there.